src/vdaf/mastic.rs

```diff
@@ -379,32 +382,418 @@ where
+            helper_share.szk_query_share_opt,
+        ) {
+            (Some(leader_query_share), Some(helper_query_share)) => Ok(self
+                .szk
+                .merge_verifiers(leader_query_share, helper_query_share)?),
+            (None, None) => Ok(None),
+            (_, _) => Err(VdafError::Uncategorized(
+                "Only one of leader and helper query shares is present".to_string(),
+            )),
+        }
+    }
+
+    fn prepare_next(
+        &self,
+        state: MasticPrepareState<T::Field, SEED_SIZE>,
+        input: MasticPrepareMessage<SEED_SIZE>,
+    ) -> Result<PrepareTransition<Self, SEED_SIZE, NONCE_SIZE>, VdafError> {
+        match (state, input) {
+            (
+                MasticPrepareState {
+                    output_shares,
+                    szk_query_state: _,
+                    verifier_len: _,
+                },
+                None,
+            ) => Ok(PrepareTransition::Finish(output_shares)),
+            (
+                MasticPrepareState {
+                    output_shares,
+                    szk_query_state,
+                    verifier_len: _,
+                },
+                Some(ref joint_share),
+            ) => {
+                self.szk.decide(szk_query_state, joint_share)?;
+                Ok(PrepareTransition::Finish(output_shares))
+            }
+        }
+    }
+
+    fn aggregate<M: IntoIterator<Item = Self::OutputShare>>(
+        &self,
+        agg_param: &MasticAggregationParam,
+        output_shares: M,
+    ) -> Result<MasticAggregateShare<T::Field>, VdafError> {
+        let mut agg_share = MasticAggregateShare::<T::Field>::from(vec![
+            T::Field::zero();
+            self.vidpf.weight_parameter
+                * agg_param
+                    .level_and_prefixes
+                    .prefixes()
+                    .len()
+        ]);
+        for output_share in output_shares.into_iter() {
+            agg_share.accumulate(&output_share)?;
+        }
+        Ok(agg_share)
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Collector for Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn unshard<M: IntoIterator<Item = Self::AggregateShare>>(
+        &self,
+        agg_param: &MasticAggregationParam,
+        agg_shares: M,
+        _num_measurements: usize,
+    ) -> Result<Self::AggregateResult, VdafError> {
+        let n = agg_param.level_and_prefixes.prefixes().len();
+        let mut agg_final = MasticAggregateShare::<T::Field>::from(vec![
+            T::Field::zero();
+            self.vidpf.weight_parameter
+                * n
+        ]);
+        for agg_share in agg_shares.into_iter() {
+            agg_final.merge(&agg_share)?;
+        }
+        let mut result = Vec::<T::AggregateResult>::with_capacity(n);
+        for i in 0..n {
+            let encoded_result = &agg_final.0
+                [i * self.vidpf.weight_parameter..(i + 1) * self.vidpf.weight_parameter];
```

Comment:
Use `chunks()` to get an iterator instead: https://doc.rust-lang.org/std/slice/struct.Chunks.html#example

This is easier to read and allows the compiler to do some more optimization.